### PR TITLE
feat: Add timeout to the subgraph fetches

### PR DESCRIPTION
**Pull request by cyaiox**
This PR updates the query fetch logic, adding a timeout to cancel the request when it takes too much time to resolve

### CODE REVIEW

File: src/ethereum/api/BaseGraphAPI.ts [hunk 1]
```diff
@@ -62,6 +64,9 @@ export class BaseGraphAPI {
       const result = await this.client.query<T, TVariables>(options)
       return result
     } catch (error) {
+      if (isErrorWithMessage(error) && error.message.includes('aborted')) {
+        throw new HTTPError(error.message, {}, STATUS_CODES.error)
+      }
```
Comment (LautaroPetaccio) [inline src/ethereum/api/BaseGraphAPI.ts:69]: Love the change! It's super important that we propagate the error when this happens.

File: src/ethereum/api/BaseGraphAPI.ts [hunk 2]
```diff
@@ -6,6 +6,8 @@ import {
 } from '@apollo/client/core'
 import { ApolloClient, NormalizedCacheObject } from '@apollo/client/core'
 import { createClient } from './graphClient'
+import { HTTPError, STATUS_CODES } from '../../common/HTTPError'
+import { isErrorWithMessage } from '../../utils/errors'
```
Comment (LautaroPetaccio) [inline src/ethereum/api/BaseGraphAPI.ts:10]: Would you mind re-ordering the imports?

File: .env.example [hunk 3]
```diff
@@ -26,6 +26,7 @@ BUILDER_SERVER_URL=
 BUILDER_SHARE_URL=
 
 COLLECTIONS_GRAPH_URL=
+GRAPH_QUERY_TIMEOUT=
```
Comment (LautaroPetaccio) [inline .env.example:29]: Did you change the builder's definition file to add the environment variable as well?

File: src/ethereum/api/collection.ts [hunk 4]
```diff
@@ -184,9 +184,14 @@ export class CollectionAPI extends BaseGraphAPI {
...
-      query: getCollectionsQuery(filters),
-    })
+    return logExecutionTime(
+      () =>
+        this.paginate(['collections'], {
+          query: getCollectionsQuery(filters),
+        }),
+      this.logger,
+      'Collections fetch'
+    )
```
Comment (LautaroPetaccio) [inline src/ethereum/api/collection.ts:194]: Do we want to keep logging the time?

File: src/ethereum/api/graphClient.ts [hunk 5]
```diff
@@ -18,3 +22,35 @@ export function createClient(url: string) {
...
+
+    clearTimeout(id)
+
+    return response
+  } catch (error) {
+    if (isErrorWithMessage(error) && error.message.includes('aborted')) {
+      const requestBody = JSON.parse(options.body as string)
+      logger.error(
+        `Timeout fetching the subgraph: ${uri}/${requestBody.operationName}`
+      )
```
Comment (LautaroPetaccio) [inline src/ethereum/api/graphClient.ts:50]: Great job! Logging them will help us a lot identify the timeouts.

File: src/ethereum/api/graphClient.ts [hunk 6]
```diff
@@ -18,3 +22,35 @@ export function createClient(url: string) {
...
+      throw new Error(error.message)
```
Comment (LautaroPetaccio) [inline src/ethereum/api/graphClient.ts:51]: LOL, creating a new error fixed the issue?

File: src/ethereum/api/graphClient.ts [hunk 7]
```diff
@@ -1,9 +1,13 @@
+import nodefetch, { RequestInit } from 'node-fetch'
 import { ApolloClient, HttpLink, InMemoryCache } from '@apollo/client/core'
-import 'isomorphic-fetch'
+import { createConsoleLogComponent } from '@well-known-components/logger'
+import { env } from 'decentraland-commons'
+import { isErrorWithMessage } from '../../utils/errors'
```
Comment (LautaroPetaccio) [inline src/ethereum/api/graphClient.ts:5]: Would you mind re-ordering this import?

File: src/ethereum/api/graphClient.ts [hunk 8]
```diff
@@ -1,9 +1,13 @@
+import nodefetch, { RequestInit } from 'node-fetch'
```
Comment (LautaroPetaccio) [inline src/ethereum/api/graphClient.ts:1]: Would you mind checking out if there's any other place where isomorphic fetch is used so we can remove it now? Thanks to this change, we no longer need the global fetch.
We can also try removing the `dom` keyword from the libraries section.

**Review by LautaroPetaccio — CHANGES_REQUESTED** [review]
Great change! I've left some small comments to discuss

File: .env.example (hunk 3 above)
Comment (LautaroPetaccio) [reply .env.example:29]: My bad, already saw it.

File: src/ethereum/api/collection.ts (hunk 4 above)
Comment (cyaiox) [reply src/ethereum/api/collection.ts:194]: not really, we're already logging the graph's queries timeout. I'll remove this

File: src/ethereum/api/graphClient.ts (hunk 6 above)
Comment (cyaiox) [reply src/ethereum/api/graphClient.ts:51]: yeap, for some reason, the aborted error wasn't captured by the apollo client :/

File: package.json [hunk 9]
```diff
@@ -62,7 +62,6 @@
     "graphql": "^15.5.0",
     "interface-datastore": "^0.7.0",
     "ipfs-unixfs-engine": "^0.35.4",
-    "isomorphic-fetch": "^2.2.1",
```
Comment (LautaroPetaccio) [inline package.json:65]: 👏

File: src/ethereum/api/BaseGraphAPI.ts (hunk 2 above)
Comment (LautaroPetaccio) [reply src/ethereum/api/BaseGraphAPI.ts:10]: They're ordered 🤦

File: src/ethereum/api/graphClient.ts (hunk 7 above)
Comment (LautaroPetaccio) [reply src/ethereum/api/graphClient.ts:5]: It was ordered, my bad 🤦

**Review by LautaroPetaccio — APPROVED** [review]
Send it 🚀

**Pull request merged**